runtime/runtimetools/perfscripts/perf-hottest

```diff
@@ -0,0 +1,177 @@
+ # [2] http://openjdk.java.net/legal/assembly-exception.html
+ #
+ # SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+# ---------------------------------------------------------------------------
+
+# This script assumes the perf.data is in the folder where the script is;
+# This script also assumes perf-$pid.map files exist in /tmp
+# There is a known bug with the perf tool that make this script not usable. Details: https://bugzilla.kernel.org/show_bug.cgi?id=203811
+
+from __future__ import print_function
+from collections import defaultdict, namedtuple
+import re, sys, textwrap
+
+try:
+    from builtins import hex as _builtin_hex
+except ImportError:
+    from __builtin__ import hex as _builtin_hex
+
+Sample = namedtuple('Sample', 'pc sym offs so thread')
+
+def main(args):
+    if len(args) == 0:
+        cmdline_error()
+
+    if args[0] in ('-h', '--help'):
+        if len(args) == 1:
+            usage(out=sys.stdout, status=0)
+        else:
+            cmdline_error()
+
+    attrs = set('pc sym offs so thread'.split())
+    group_by_attr = args[0]
+    if group_by_attr not in attrs:
+        cmdline_error()
+
+    pred = lambda s: True
+    def pred_compose(prev, attr, attr_re):
+        return lambda s: (prev(s) and attr_re.search(str(getattr(s, attr))))
+
+    for arg in args[1:]:
+        if '=' not in arg:
+            printerr('error: filter argument not of the form <attr>=<regex>:', arg)
+            cmdline_error()
+
+        cur_filter_attr, cur_filter_re_src = arg.split('=', 1)
+        try:
+            cur_filter_re = re.compile(cur_filter_re_src)
+        except re.error:
+            printerr('error: invalid regular expression:', cur_filter_re_src)
+            cmdline_error()
+
+        pred = pred_compose(pred, cur_filter_attr, cur_filter_re)
+
+    samples = (s for s in read_samples(sys.stdin) if pred(s))
+    counts = counts_by(group_by_attr, samples)
+    total_count = sum(n for key, n in counts.iteritems())
+    assert (total_count == 0) == (len(counts) == 0)
+
+    if len(counts) > 0:
+        table = []
+        for key, n in sorted(counts.iteritems(), key=lambda pair: pair[1]):
+            pct = 100.0 * float(n) / float(total_count)
+            table.append((n, '{0:6.2f}%'.format(pct), key))
+
+        hex_key_fmt = '-0{width}x'
+        key_fmt = ''
+        if group_by_attr == 'pc':
+            key_fmt = hex_key_fmt
+        elif group_by_attr == 'offs' and '?' not in counts:
+            key_fmt = hex_key_fmt
+        format_table(('-{width}', '', key_fmt), table)
+
+    print('', total_count, '        --total--')
+
+def cmdline_error():
+    usage(out=sys.stderr, status=1)
+
+def printerr(*args):
+    print('error:', *args, out=sys.stderr)
+
+def usage(out, status):
+    msg = textwrap.dedent('''\
+        usage: perf script | perf-hottest <attr> [<attr>=<regex>...]
+        Attributes: pc, sym, offs, so, thread
```

Comment:
Can this also list what these attributes mean?

Reply:
Will add more details